packages/devextreme-react/src/core/template-manager.tsx

```diff
@@ -0,0 +1,180 @@
+import * as React from 'react';
+import * as events from 'devextreme/events';
+
+import {
+  useState,
+  useMemo,
+  useCallback,
+  useRef,
+  FC,
+} from 'react';
+
+import {
+  TemplateManagerProps,
+  GetRenderFuncFn,
+  DXTemplateCollection,
+  TemplateFunc,
+} from './types';
+
+import TemplateWrapper from './template-wrapper';
+import { RenderedTemplateInstances, generateID } from './helpers';
+import { DX_REMOVE_EVENT } from './component-base';
+import { ITemplateArgs } from './template';
+import { getOption as getConfigOption } from './config';
+import { ITemplate } from './configuration/config-node';
+
+function normalizeProps(props: ITemplateArgs): ITemplateArgs | ITemplateArgs['data'] {
+  if (getConfigOption('useLegacyTemplateEngine')) {
+    const model = props.data;
+    if (model && Object.prototype.hasOwnProperty.call(model, 'key')) {
+      model.dxkey = model.key;
+    }
+    return model;
+  }
+  return props;
+}
+
+export const TemplateManager: FC<TemplateManagerProps> = ({ init }) => {
+  const [renderedInstances, setRenderedInstances] = useState(new RenderedTemplateInstances());
```

Comment:
They are not rendered, since they are not connected to DOM/VDOM. These are just data-transfer objects (`TemplateInstanceDefinition`).

```suggestion
  const [instantiationModels, setInstantiationModels] = useState(new RenderedTemplateInstances());
```